omisego-sdk/src/main/java/co/omisego/omisego/security/KeyManagerMarshmallow.kt

```diff
@@ -20,53 +19,48 @@ import javax.crypto.spec.GCMParameterSpec
+    }
+    override val decryptCipher: Cipher by lazy {
+        Cipher.getInstance(AES_MODE).also { initCipher(it, Cipher.DECRYPT_MODE) }
     }
 
-    // Generate key function for Android version 6.0 Marshmallow or above
-    @RequiresApi(Build.VERSION_CODES.M)
-    fun create() {
-        keyStore = KeyStore.getInstance(ANDROID_KEY_STORE)
-        keyStore.load(null)
-        if (!keyStore.containsAlias(keyAlias)) {
-            val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEY_STORE)
-
-            val spec = KeyGenParameterSpec.Builder(keyAlias, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
-                    .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
-                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
-                    .setRandomizedEncryptionRequired(false)
-                    .build()
+    private val secretKey: Key
+        get() = keyStore.getKey(keyAlias, null)
+    private val gcmParameterSpec: GCMParameterSpec
+        get() = GCMParameterSpec(128, iv.toByteArray())
 
-            keyGenerator.init(spec)
-            keyGenerator.generateKey()
-        }
+    companion object {
+        const val AES_MODE = "AES/GCM/NoPadding"
     }
 
-    fun setIV(iv: String) {
-        this.iv = iv
-    }
+    // Generate key function for Android version 6.0 Marshmallow or above
+    override fun generateKey(context: Context) {
+        val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES,
+                                                    ANDROID_KEY_STORE)
 
-    private fun getSecretKey(): Key = keyStore.getKey(keyAlias, null)
+        val spec = KeyGenParameterSpec.Builder(keyAlias,
+                                               KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
+                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+                .setRandomizedEncryptionRequired(false)
+                .build()
 
-    // Encrypt data
-    fun encrypt(input: ByteArray): String {
-        val cipher = Cipher.getInstance(AES_MODE)
-        cipher.init(Cipher.ENCRYPT_MODE, getSecretKey(), GCMParameterSpec(128, iv.toByteArray()))
-        val encodedBytes: ByteArray = cipher.doFinal(input)
-        return Base64.encodeToString(encodedBytes, Base64.DEFAULT)
+        keyGenerator.init(spec)
+        keyGenerator.generateKey()
     }
 
-    // Decrypt data
-    fun decrypt(encrypted: ByteArray): String {
-        val c = Cipher.getInstance(AES_MODE)
-        c.init(Cipher.DECRYPT_MODE, getSecretKey(), GCMParameterSpec(128, iv.toByteArray()))
-        val decoded = Base64.decode(encrypted, Base64.DEFAULT)
-        val decodedBytes = c.doFinal(decoded)
-        return String(decodedBytes)
+    override fun initCipher(cipher: Cipher, opmode: Int) {
+        if (opmode !in arrayOf(Cipher.ENCRYPT_MODE, Cipher.ENCRYPT_MODE)) {
```

Comment:
Instead of
```kotlin
if (opmode !in arrayOf(Cipher.ENCRYPT_MODE, Cipher.ENCRYPT_MODE)) {
```
it should be
```kotlin
if (opmode !in arrayOf(Cipher.ENCRYPT_MODE, Cipher.DECRYPT_MODE)) {
```